checkasm/checkasm.c

```diff
@@ -0,0 +1,924 @@
+    case EXCEPTION_ILLEGAL_INSTRUCTION:
+    case EXCEPTION_PRIV_INSTRUCTION:
+        err = "illegal instruction";
+        break;
+    case EXCEPTION_ACCESS_VIOLATION:
+    case EXCEPTION_ARRAY_BOUNDS_EXCEEDED:
+    case EXCEPTION_DATATYPE_MISALIGNMENT:
+    case EXCEPTION_IN_PAGE_ERROR:
+    case EXCEPTION_STACK_OVERFLOW:
+        err = "segmentation fault";
+        break;
+    default:
+        return EXCEPTION_CONTINUE_SEARCH;
+    }
+    state.catch_signals = 0;
+    checkasm_fail_func(err);
+    checkasm_load_context();
+    return EXCEPTION_CONTINUE_EXECUTION; /* never reached, but shuts up gcc */
+}
+#else
+static void signal_handler(const int s) {
+    if (state.catch_signals) {
+        state.catch_signals = 0;
+        checkasm_fail_func(s == SIGFPE ? "fatal arithmetic error" :
+                           s == SIGILL ? "illegal instruction" :
+                                         "segmentation fault");
+        checkasm_load_context();
+    } else {
+        /* fall back to the default signal handler */
+        static const struct sigaction default_sa = { .sa_handler = SIG_DFL };
+        sigaction(s, &default_sa, NULL);
+        raise(s);
+    }
+}
+#endif
+
+/* Compares a string with a wildcard pattern. */
+static int wildstrcmp(const char *str, const char *pattern) {
+    const char *wild = strchr(pattern, '*');
+    if (wild) {
+        const size_t len = wild - pattern;
+        if (strncmp(str, pattern, len)) return 1;
+        while (*++wild == '*');
+        if (!*wild) return 0;
+        str += len;
+        while (*str && wildstrcmp(str, wild)) str++;
+        return !*str;
+    }
+    return strcmp(str, pattern);
+}
+
+/* Perform tests and benchmarks for the specified
+ * cpu flag if supported by the host */
+static void check_cpu_flag(const char *const name, unsigned flag) {
+    const unsigned old_cpu_flag = state.cpu_flag;
+
+    flag |= old_cpu_flag;
+    dav1d_set_cpu_flags_mask(flag);
+    state.cpu_flag = dav1d_get_cpu_flags();
+
+    if (!flag || state.cpu_flag != old_cpu_flag) {
+        state.cpu_flag_name = name;
+        for (int i = 0; tests[i].func; i++) {
+            if (state.test_pattern && wildstrcmp(tests[i].name, state.test_pattern))
+                continue;
+            xor128_srand(state.seed);
+            state.current_test_name = tests[i].name;
+            tests[i].func();
+        }
+    }
+}
+
+/* Print the name of the current CPU flag, but only do it once */
+static void print_cpu_name(void) {
+    if (state.cpu_flag_name) {
+        color_printf(COLOR_YELLOW, "%s:\n", state.cpu_flag_name);
+        state.cpu_flag_name = NULL;
+    }
+}
+
+static unsigned get_seed(void) {
+#ifdef _WIN32
+    LARGE_INTEGER i;
+    QueryPerformanceCounter(&i);
+    return i.LowPart;
+#elif defined(__APPLE__)
+    return (unsigned) mach_absolute_time();
+#else
+    struct timespec ts;
+    clock_gettime(CLOCK_MONOTONIC, &ts);
+    return (unsigned) (1000000000ULL * ts.tv_sec + ts.tv_nsec);
+#endif
```

Comment:
I know this is copied from dav1d, but do you know why `__APPLE__` isn't also using `clock_gettime` here? `clock_gettime`  is part of POSIX since 1993, so it should also exist on macOS.

Reply:
No, it doesn’t exit on macOS.

POSIX says:

> These functions are part of the Timers option and need not be available on all implementations.

Reply:
Oh, I missed this being optional. Thanks!

Modern illumos [documents](https://illumos.org/man/3C/clock_gettime) `CLOCK_MONOTONIC` as being provided as an alias for `CLOCK_HIGHRES`. And sure enought checking the [current headers](https://github.com/illumos/illumos-gate/blob/9ecd05bdc59e4a1091c51ce68cce2028d5ba6fd1/usr/src/uts/common/sys/time_impl.h#L124-L126) both names are available (though the alias goes in the other direction).

Reply:
`clock_gettime` has been present on macOS since 10.12. Apple suggests replacing `mach_absolute_time` with `CLOCK_UPTIME_RAW`, which is also preferable over `CLOCK_MONOTONIC` for this use-case when available, since it isn't affected by NTP-driven rate changes (both are unaffected by NTP-driven time _jumps_). Specifically, Apple recommends using `clock_gettime_nsec_np(CLOCK_UPTIME_RAW)`, which is slightly simpler to handle.

On macOS, `CLOCK_MONOTONIC`, `CLOCK_MONOTONIC_RAW`, and `CLOCK_UPTIME_RAW` are implemented in userspace, on top of `_mach_boottime_usec`, `mach_continuous_time`, and `mach_absolute_time` respectively. `mach_continuous_time` is just `mach_absolute_time`, except that it counts while the system is asleep, which is irrelevant to us. So, `mach_absolute_time` is probably preferable for our use-case, since we don't care about the actual time-base (optimally we'd be counting _cycles_, there's just no documented/unprivileged non-x86-specific way to do that on macOS), and `clock_gettime_nsec_np(CLOCK_MONOTONIC_RAW)` and `clock_gettime_nsec_np (CLOCK_UPTIME_RAW)` have to call their underlying functions and then perform a time-base conversion.